### DISCUSSION TITLE
Transfer data to the sketch from custom pages

### DISCUSSION THREAD

**Discussion by everthu** [Hieromon/AutoConnect#364] [How can I?] [answered]
Hi,
I like to make a custom page including colorpicker to control my ws2812 led strip.
Therefor I made a page including a link to a jscolor javascript link to show the colorpicker and grab the selected color and put it in the URL bar. 
In [this](https://randomnerdtutorials.com/esp32-esp8266-rgb-led-strip-web-server/) example the selected color is put as a request on the URL bar which is fetched somewhere else in the sketch.
With a custom page within autoconnect I am able to show the colorpicker and select a color and it's shown as expected at the URL, however I have no idea how to fetch the data somewhere in the loop of my sketch.
Does anyone give me some hints about how to do this, or is there another possibility to transfer the data from the html page to the sketch?

Regards Evert Huijben

**Comment by Hieromon** [comment, marked as answer]
What you want to know is already in the documentation.
https://hieromon.github.io/AutoConnect/achandling.html#custom-field-data-handling
https://hieromon.github.io/AutoConnect/achandling.html#where-to-pick-up-the-values

But if you try to apply your referenced demonstration sketch directly to AutoConnect, you will fail. The WebServer that the demonstration sketch envisions is neither the ESP32 WebServer class nor the ESP8266WebServer class. Your referred sketch is an example when using Arduino's WebServer class.
The WebServer class in the Arduino standard library is different from the ESP8266WebServer or ESP32 WebServer class. The ESP32 WebServer class includes and extends the Arduino WebServer class, and AutoConnect's custom Web page handler takes advantage of that extended functionality.
AutoConnect's custom Web page handler is compliant with the ESP8266WebServer/ESP32WebServer class page request handler, Requests from clients are processed by *AutoConnect::handleClient* function. It cannot be handled directly in the sketch loop. So all client requests are handled by the page request handlers. Not in the `loop` function.

### Now, let's replace the sketch code you referred to with AutoConnect.

First is from HTML. The HTML below is generated by the sketch you referred to.

```html
<!DOCTYPE html>
<html>
<head>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="icon" href="data:,">
    <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css">
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jscolor/2.0.4/jscolor.min.js"></script>
</head>
<body>
    <div class="container">
        <div class="row">
            <h1>ESP Color Picker</h1>
        </div>
        <a class="btn btn-primary btn-lg" href="#" id="change_color" role="button">Change Color</a> 
        <input class="jscolor {onFineChange:'update(this)'}" id="rgb">
    </div>
    <script>
        function update(picker) {
            document.getElementById('rgb').innerHTML = Math.round(picker.rgb[0]) + ', ' +  Math.round(picker.rgb[1]) + ', ' + Math.round(picker.rgb[2]);
            document.getElementById("change_color").href="?r" + Math.round(picker.rgb[0]) + "g" +  Math.round(picker.rgb[1]) + "b" + Math.round(picker.rgb[2]) + "&";
        }
    </script>
</body>
</html>
```

It's simple and has only a color picker and a button to send the RBG value to the server. The method of sending to the server has left to the *jscolor* plugin.
And the HTTP request destination (URL) is the root path of the server. As `r255g255b255&` where the argument of the query string indicates the RGB value. The HTTP request with RGB value contained query string is sent to the server by manipulating the `href` attribute of the `<a>` tag of `id=change_color`. It is `document.getElementById("change_color").href="?r" + Math.round(picker.rgb[0]) + "g" +  Math.round(picker.rgb[1]) + "b" + Math.round(picker.rgb[2]) + "&";`.

We have two options for replacing this page with an AutoConnect custom web page. After all, the option is "What to do with page transitions?"

#### 1. Replace the sample sketch simply

The color value selected with the color picker is retrieved by the [custom web page handler](https://hieromon.github.io/AutoConnect/api.html#on) during the page transition. In the sample sketch, the page transition destination (URL) by the `id="chnage_color"` link is the root path. And the page content returned by the server is the same. In other words, the same page content is repeated. In the AutoConnectAux definition, write as follows:

```json
{
  "title": "Color Picker",
  "uri": "/",
  "menu": true,
  "element": [
    {
      "name": "jscolor",
      "type": "ACElement",
      "value": "<script src=\"https://cdnjs.cloudflare.com/ajax/libs/jscolor/2.0.4/jscolor.min.js\"></script>"
    },
    {
      "name": "row", 
      "type": "ACElement",
      "value": "<h1>ESP Color Picker</h1>"
    },
    {
      "name": "change_color",
      "type": "ACElement",
      "value": "<a class=\"btn btn-primary btn-lg\" href=\"#\" id=\"change_color\" role=\"button\">Change Color</a><input class=\"jscolor{onFineChange:'update(this)'}\" id=\"rgb\">"
    },
    {
      "name": "js",
      "type": "ACElement",
      "value": "<script>function update(picker){document.getElementById('rgb').innerHTML=Math.round(picker.rgb[0])+', '+Math.round(picker.rgb[1])+', '+Math.round(picker.rgb[2]);document.getElementById('change_color').href=\"?r\"+Math.round(picker.rgb[0])+\"g\"+Math.round(picker.rgb[1])+\"b\"+Math.round(picker.rgb[2])+\"&\";}</script>"
    }
  ]
}
```

The query string as `r255g255b255&` contained in the URL of the HTTP request can be [retrieved from the PageArgument](https://hieromon.github.io/AutoConnect/achandling.html#how-you-can-reach-the-values) passed to AutoConnect's custom web page handler. The [PageArgument](https://github.com/Hieromon/PageBuilder#arguments-of-invoked-user-function) class is compatible with the [arg of the ESP8266WebServer class](https://github.com/esp8266/Arduino/tree/master/libraries/ESP8266WebServer#advanced-options). Normally, the query string takes formed `name=value`, but if it is a simple string like `r255g255b255&`, it will result in a parsing error as a parameter. 

```javascript
document.getElementById('change_color').href="?r255g255b255&"
```

To avoid parsing error, change the query string to the `name=value` format.

```javascript
document.getElementById('change_color').href="/?color=r255g255b255"
```

So the code to get the color picker value using the PageArgument's arg function inside a custom web page handler looks like this, according to the [sketch example you referenced](https://randomnerdtutorials.com/esp32-esp8266-rgb-led-strip-web-server/):

```cpp
String colorPicker(AutoConnectAux& aux, PageArgument& args) {
  if (args.hasArg("color")) {
    String  rgb = args.arg("color");
    int pos1 = rgb.indexOf('r');
    int pos2 = rgb.indexOf('g');
    int pos3 = rgb.indexOf('b');
    int pos4 = rgb.length();
    r = rgb.substring(pos1 + 1, pos2).toInt();
    g = rgb.substring(pos2 + 1, pos3).toInt();
    b = rgb.substring(pos3 + 1, pos4).toInt();
  }
  return String();
}
```

The next thing to do is to PWM output in the loop function with the RGB values retrieved by the custom web page handler. Only a custom web page handler can detect if the RBG value has changed. Therefore, the flag that notifies the color change is shared globally with the loop function.

```cpp
AutoConnect portal;

// RGB value
int r;
int g;
int b;

// Request for color change
bool colorChange;

String colorPicker(AutoConnectAux& aux, PageArgument& args) {
  if (args.hasArg("color")) {
    String  rgb = args.arg("color");
    int pos1 = rgb.indexOf('r');
    int pos2 = rgb.indexOf('g');
    int pos3 = rgb.indexOf('b');
    int pos4 = rgb.length();
    r = rgb.substring(pos1 + 1, pos2).toInt();
    g = rgb.substring(pos2 + 1, pos3).toInt();
    b = rgb.substring(pos3 + 1, pos4).toInt();
    // Notify the color change
    colorChange = true;
  }
  return String();
}
...
void loop() {
  // 
  if (colorChange) {
    ledcWrite(redChannel, r);
    ledcWrite(greenChannel, g);
    ledcWrite(blueChannel, b);
    Serial.printf("R[%d] G[%d] B[%d]\n", r, g, b);
    colorChange = false; // Reset
  }
  portal.handleClient();
}
```

A sketch with AutoConnect applied in this way would look like this (a case for ESP32):

```cpp
#include <Arduino.h>
#include <WiFi.h>
#include <WebServer.h>
#include <AutoConnect.h>

const char COLORPICKER[] PROGMEM = R"(
{
  "title": "Color Picker",
  "uri": "/",
  "menu": true,
  "element": [
    {
      "name": "jscolor",
      "type": "ACElement",
      "value": "<script src=\"https://cdnjs.cloudflare.com/ajax/libs/jscolor/2.0.4/jscolor.min.js\"></script>"
    },
    {
      "name": "row", 
      "type": "ACElement",
      "value": "<h1>ESP Color Picker</h1>"
    },
    {
      "name": "change_color",
      "type": "ACElement",
      "value": "<a class=\"btn btn-primary btn-lg\" href=\"#\" id=\"change_color\" role=\"button\">Change Color</a><input class=\"jscolor{onFineChange:'update(this)'}\" id=\"rgb\">"
    },
    {
      "name": "js",
      "type": "ACElement",
      "value": "<script>function update(picker){document.getElementById('rgb').innerHTML=Math.round(picker.rgb[0])+', '+Math.round(picker.rgb[1])+', '+Math.round(picker.rgb[2]);document.getElementById('change_color').href=\"?color=r\"+Math.round(picker.rgb[0])+\"g\"+Math.round(picker.rgb[1])+\"b\"+Math.round(picker.rgb[2]);}</script>"
    }
  ]
}
)";

AutoConnect portal;

// Red, green, and blue pins for PWM control
const int redPin = 13;     // 13 corresponds to GPIO13
const int greenPin = 12;   // 12 corresponds to GPIO12
const int bluePin = 14;    // 14 corresponds to GPIO14

// Setting PWM frequency, channels and bit resolution
const int freq = 5000;
const int redChannel = 0;
const int greenChannel = 1;
const int blueChannel = 2;
// Bit resolution 2^8 = 256
const int resolution = 8;

// RGB value
int r = 0;
int g = 0;
int b = 0;

// Request for color change
bool colorChange;

String colorPicker(AutoConnectAux& aux, PageArgument& args) {
  if (args.hasArg("color")) {
    String  rgb = args.arg("color");
    int pos1 = rgb.indexOf('r');
    int pos2 = rgb.indexOf('g');
    int pos3 = rgb.indexOf('b');
    int pos4 = rgb.length();
    r = rgb.substring(pos1 + 1, pos2).toInt();
    g = rgb.substring(pos2 + 1, pos3).toInt();
    b = rgb.substring(pos3 + 1, pos4).toInt();
    colorChange = true;
  }
  return String();
}

void setup() {
  delay(1000);
  Serial.begin(115200);
  Serial.println();

  // configure LED PWM functionalitites
  ledcSetup(redChannel, freq, resolution);
  ledcSetup(greenChannel, freq, resolution);
  ledcSetup(blueChannel, freq, resolution);
  
  // attach the channel to the GPIO to be controlled
  ledcAttachPin(redPin, redChannel);
  ledcAttachPin(greenPin, greenChannel);
  ledcAttachPin(bluePin, blueChannel);
  colorChange = true;

  portal.load(COLORPICKER);
  portal.on("/", colorPicker);
  portal.begin();
}

void loop() {
  if (colorChange) {
    Serial.printf("R[%d] G[%d] B[%d]\n", r, g, b);
    colorChange = false;
  }
  ledcWrite(redChannel, r);
  ledcWrite(greenChannel, g);
  ledcWrite(blueChannel, b);

  portal.handleClient();
}
```

#### 2. Migrate to using [XHR](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/Using_XMLHttpRequest)

The sketch you referenced uses the old method of HTTP requests. It puts a load on the traffic with page transitions and rewrites. The above sketch does not require page transitions as there is a web page for color RGB value manipulation. All you have to do is have a dialog.
Custom web pages can use [XMLHttpRequest(XHR)](https://developer.mozilla.org/ja/docs/Web/API/XMLHttpRequest) to perform sketching and HTTP communication without page transitions. There is a rough way to implement it in the [documentation](https://hieromon.github.io/AutoConnect/achandling.html#communicate-with-the-sketch-using-xhr), but I think your questions will be useful to other users as well, so here I'll try to implement a custom web page for AutoConnect using XHR.

The first thing to consider is a break from the *jscolor*. The *jscolor* is a great plugin with good usability, but even the mini version reaches 60 Kb in size, which makes it impractical to include it in your sketches.
But we have another option. The [`input type="color"`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/color) tag natively supported by HTML doesn't have as many features as the *jscolor*, but it does have enough user interface to select colors. And you can use [AutoConnectElement](https://hieromon.github.io/AutoConnect/achandling.html#place-html-elements-undefined-in-autoconnectelements) to embed the `input type="color"` tag in your custom web page.

The HTML description for embedding the `input type="color"` tag in the HTML page is as follows, for example:

```html
<input type="color" name="colorWell" id="colorWell" value="#e66465">
```

You can replace this HTML with a JSON description to place it on your AutoConnect custom web page:

```json
{
  "name": "colorWell",
  "type": "ACElement",
  "value": "<input type=\"color\" name=\"colorWell\" id=\"colorWell\" value=\"#e66465\">"
}
```

The next step is to arrange the text box for input and output of color values. Its text box allows you to work with the color selection operation in the color picker.

```html
<input type="text" name="rgbValue" id="rgbValue" value="#e66465">
<input type="color" name="colorWell" id="colorWell" value="#e66465">

```

To ensure that the values of the above two tags change automatically as you operate, use the [event notifications](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/change_event) provided by those elements.
The [event listener](https://developer.mozilla.org/en-US/docs/Web/API/EventListener) is an interface for detecting and dealing with events that occur in an element, and we can usually implement it in JavaScript. Then, register the implemented JavaScript event listener function via the [addEventListener](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener) method.

```javascript
var rgbInput;
var colorWell;

function pageSetup() {
    colorWell = document.getElementById("colorWell");
    colorWell.addEventListener("input", updateColor, false);
    colorWell.addEventListener("change", updateColor, false);
    colorWell.select();
    rgbInput = document.getElementById("rgbValue");
    rgbInput.addEventListener("change", setColor, true);
}

function setColor(e) {
    reflectText(e.target.value);
    colorWell.value = e.target.value;
}

function updateColor(e) {
    rgbInput.value = e.target.value;
    reflectText(e.target.value);
}

function reflectText(color) {
    rgbInput.style.backgroundColor = color;
    var rgb = hexToRgb(color);
    var textColor = oppositeColor(rgb["r"], rgb["g"], rgb["b"]);
    rgbInput.style.color = textColor;
}

function hexToRgb(hex) {
    var result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
    return result ? {
        r: parseInt(result[1], 16),
        g: parseInt(result[2], 16),
        b: parseInt(result[3], 16)
    } : null;
}

function oppositeColor(R, G, B) {
    if(!isNaN(R + G + B) && 0 <= R && R <=255 && 0 <= G && G <=255 && 0 <= B && B <=255) {
        var comColor = "rgb(" + (255 - R) + ", " + (255 - G) + ", " + (255 - B) + ")";
        return comColor;
    } else {
        return null;
    }
}
```

And finally the implementation of the XHR part. This is very easy to implement, contrary to the image you get from words.
However, there is one thing to be aware of. **The XHR request URL should be separate from the custom web page URL**. The XHR request URL is itself an endpoint for receiving different requests. If the URL of the XHR endpoint and the page itself is the same, the page will not be displayed. In this demonstration, let's assign the URL of the XHR endpoint for color change to `/change`. 

```javascript
var xhr;
xhr = new XMLHttpRequest();
xhr.open("GET", "/change?color=" + "r" + rgb["r"] + "g" + rgb["g"] + "b" + rgb["b"]);
xhr.responseType = "text";
xhr.send();
```

Only above one. This script will send a query string with `color=r255g255b255` to the URL:`/change` endpoint.

Combining all the above parts, the HTML code to send to the client should look like this:

```html
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>RGB Color Picker</title>
</head>
<body>
    <h3>ESP Color Picker</h3>
    <div>
        <button type="button" name="change" id="change" onclick="changeColor()" value="Change Color">Change Color</button>
        <input type="text" name="rgbValue" id="rgbValue" value="#e66465">
        <input type="color" name="colorWell" id="colorWell" value="#e66465">
    </div>
    <script>
        var xhr;
        var rgbInput;
        var colorWell;
        function pageSetup() {
            xhr = new XMLHttpRequest();
            colorWell = document.getElementById("colorWell");
            colorWell.addEventListener("input", updateColor, false);
            colorWell.addEventListener("change", updateColor, false);
            colorWell.select();
            rgbInput = document.getElementById("rgbValue");
            rgbInput.addEventListener("change", setColor, true);
        }
        function setColor(e) {
            reflectText(e.target.value);
            colorWell.value = e.target.value;
        }
        function changeColor() {
            console.log(rgbInput.value);
            var rgb = hexToRgb(rgbInput.value);
            xhr.open("GET", "/change?color=" + "r" + rgb["r"] + "g" + rgb["g"] + "b" + rgb["b"]);
            xhr.responseType = "text";
            xhr.send();
        }
        function updateColor(e) {
            rgbInput.value = e.target.value;
            reflectText(e.target.value);
        }
        function reflectText(color) {
            rgbInput.style.backgroundColor = color;
            var rgb = hexToRgb(color);
            var textColor = oppositeColor(rgb["r"], rgb["g"], rgb["b"]);
            rgbInput.style.color = textColor;
        }
        function hexToRgb(hex) {
            var result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
            return result ? {
                r: parseInt(result[1], 16),
                g: parseInt(result[2], 16),
                b: parseInt(result[3], 16)
            } : null;
        }
        function oppositeColor(R, G, B) {
            if(!isNaN(R + G + B) && 0 <= R && R <=255 && 0 <= G && G <=255 && 0 <= B && B <=255) {
                var comColor = "rgb(" + (255 - R) + ", " + (255 - G) + ", " + (255 - B) + ")";
                return comColor;
            } else {
                return null;
            }
        }
        window.onload = pageSetup;
    </script>
</body>
</html>
```

This script size is about 2.5 Kb. This size can handle even the poor memory capacity of the ESP8266.

The remaining task is server-side processing. To receive the RGB values sent by XHR, use the ESP8266SWebserver or ESP32 WebServer [request handler](https://github.com/esp8266/Arduino/tree/master/libraries/ESP8266WebServer#client-request-handlers) natively to respond to the XHR request and retrieve the query string. You need to implement this procedure as a request handler function of ESP8266WebServer/ESP32WebServer class.

The logic for responding to XHR requests is very similar to the custom web page handler I wrote earlier. (Actually, the same thing to do)
There are two differences.
1. Describe as a request handler of WebServer class. Therefore, the method to register the handler becomes the *WebServer::on* function.
2. The handler sends the HTML response. (If you use AutoConnect's custom web page handler, the HTML response is sent by AutoConnect. The handler only assembles part of the content of the response page)
And in this specification, the response content is empty because the server-side only receives the changed RGB value. This means that the client browser does not have to rewrite the page.

```cpp
WebServer server;

void onColorChange() {
  if (server.hasArg("color")) {
    String  rgb = server.arg("color"); // Get Query string
    int pos1 = rgb.indexOf('r');
    int pos2 = rgb.indexOf('g');
    int pos3 = rgb.indexOf('b');
    int pos4 = rgb.length();
    r = rgb.substring(pos1 + 1, pos2).toInt();
    g = rgb.substring(pos2 + 1, pos3).toInt();
    b = rgb.substring(pos3 + 1, pos4).toInt();
    colorChange = true;
  }
  server.send(200, "text/plain", ""); // Response content is empty.
}
...
server.on("/change", onColorChange); // Register XHR request handler
```

So far, we're talking about how to implement an efficient web application rather than how to use AutoConnect. After all, it is how well you can use HTML and JavaScript.

```cpp
#include <Arduino.h>
#include <WiFi.h>
#include <WebServer.h>
#include <AutoConnect.h>

#define XHR_ENDPOINT  "/change"

const char COLORPICKER[] PROGMEM = R"(
{
  "title": "Color Picker",
  "uri": "/",
  "menu": true,
  "element": [
    {
      "name": "row",
      "type": "ACElement",
      "value": "<h1>ESP Color Picker</h1>"
    },
    {
      "name": "change",
      "type": "ACButton",
      "value": "Change Color",
      "action": "changeColor();"
    },
    {
      "name": "rgbValue",
      "type": "ACInput",
      "value": "Change Color"
    },
    {
      "name": "colorWell",
      "type": "ACElement",
      "value": "<input type=\"color\" name=\"colorWell\" id=\"colorWell\">"
    },
    {
      "name": "js",
      "type": "ACElement",
      "value": "<script>var xhr,rgbInput,colorWell;function pageSetup(){xhr=new XMLHttpRequest,(colorWell=document.getElementById(\"colorWell\")).addEventListener(\"input\",updateColor,!1),colorWell.addEventListener(\"change\",updateColor,!1),colorWell.select(),(rgbInput=document.getElementById(\"rgbValue\")).addEventListener(\"change\",setColor,!0)}function setColor(e){reflectText(e.target.value),colorWell.value=e.target.value}function changeColor(){console.log(rgbInput.value);var e=hexToRgb(rgbInput.value);xhr.open(\"GET\",\")" XHR_ENDPOINT R"(?color=r\"+e.r+\"g\"+e.g+\"b\"+e.b),xhr.responseType=\"text\",xhr.send()}function updateColor(e){rgbInput.value=e.target.value,reflectText(e.target.value)}function reflectText(e){rgbInput.style.backgroundColor=e;var t=hexToRgb(e),o=oppositeColor(t.r,t.g,t.b);rgbInput.style.color=o}function hexToRgb(e){var t=/^#?([a-f\\d]{2})([a-f\\d]{2})([a-f\\d]{2})$/i.exec(e);return t?{r:parseInt(t[1],16),g:parseInt(t[2],16),b:parseInt(t[3],16)}:null}function oppositeColor(e,t,o){return!isNaN(e+t+o)&&0<=e&&e<=255&&0<=t&&t<=255&&0<=o&&o<=255?\"rgb(\"+(255-e)+\", \"+(255-t)+\", \"+(255-o)+\")\":null}window.onload=pageSetup</script>"
    }
  ]
}
)";

WebServer   server;
AutoConnect portal(server);

// Red, green, and blue pins for PWM control
const int redPin = 13;     // 13 corresponds to GPIO13
const int greenPin = 12;   // 12 corresponds to GPIO12
const int bluePin = 14;    // 14 corresponds to GPIO14

// Setting PWM frequency, channels and bit resolution
const int freq = 5000;
const int redChannel = 0;
const int greenChannel = 1;
const int blueChannel = 2;
// Bit resolution 2^8 = 256
const int resolution = 8;

// RGB value
int r = 0;
int g = 0;
int b = 0;

// Request for color change
bool colorChange;

void onColorChange() {
  if (server.hasArg("color")) {
    String  rgb = server.arg("color");
    int pos1 = rgb.indexOf('r');
    int pos2 = rgb.indexOf('g');
    int pos3 = rgb.indexOf('b');
    int pos4 = rgb.length();
    r = rgb.substring(pos1 + 1, pos2).toInt();
    g = rgb.substring(pos2 + 1, pos3).toInt();
    b = rgb.substring(pos3 + 1, pos4).toInt();
    colorChange = true;
  }
  server.send(200, "text/plain", "");
}

void setup() {
  delay(1000);
  Serial.begin(115200);
  Serial.println();

  // configure LED PWM functionalitites
  ledcSetup(redChannel, freq, resolution);
  ledcSetup(greenChannel, freq, resolution);
  ledcSetup(blueChannel, freq, resolution);

  // attach the channel to the GPIO to be controlled
  ledcAttachPin(redPin, redChannel);
  ledcAttachPin(greenPin, greenChannel);
  ledcAttachPin(bluePin, blueChannel);
  colorChange = true;

  server.on(XHR_ENDPOINT, onColorChange); // Register XHR endpoint handler
  portal.load(COLORPICKER);
  portal.begin();
}

void loop() {
  if (colorChange) {
    Serial.printf("R[%d] G[%d] B[%d]\n", r, g, b);
    colorChange = false;
  }
  ledcWrite(redChannel, r);
  ledcWrite(greenChannel, g);
  ledcWrite(blueChannel, b);
  portal.handleClient();
}
```

### It's strange to say it by myself
If we depend on JavaScript so far, it may not make much sense to use AutoConnect.
This time around, I've tried to help other users who are struggling to implement the dynamic behavior of AutoConnect's custom web pages. I hope this topic helps.

**Comment by everthu** [comment]
Hi Hieromon,
Thank you for the extensive answer and the examples,
This is a good example for improving my autoconnect skills, especially for the dynamic part.
I've started implementing, it looks promising so far.
Thank you, Evert

**Comment by everthu** [comment]
Hi Hieromon,

Yesterday I played around with the XHR solution you proposed, this looks really nice.
I've just one question:
Is there any way to store the selected color, in a way that when I refresh the page that I can still see the selected color.
If I used the autoconnect elements, it's done automatically. I like to see the same behaviour with the color selection.
Any suggestion?

Regards, Evert Huijben
